File: src/algorithms/Remove_Duplicates_from_Sorted_List_II.py
# -*- coding: utf-8 -*-


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    # @param {ListNode} head
    # @return {ListNode}
    def deleteDuplicates(self, head):
        if head is None:
            return None
        if head.next is None:
            return head
        while True:
            cnt = 1
            while head.next and head.val == head.next.val:
                cnt += 1
                head = head.next
            if cnt > 1:
                head = head.next
            if head is None:
                return None
            if head.next is None or head.val != head.next.val:
                break
        node, pre = head.next, head
        if node:
            while True:
                cnt = 1
                while node.next and node.val == node.next.val:
                    cnt += 1
                    node = node.next
                if cnt == 1:
                    pre.next = node
                    pre = node
                node = node.next
                if not node or not node.next:
                    break
            pre.next = node
        return head


if __name__ == '__main__':
    solution = Solution()
    a, b = ListNode(1), ListNode(1)
    c, d = ListNode(1), ListNode(1)
    e = ListNode(2)
    a.next, b.next = b, c
    c.next, d.next = d, e
    node = solution.deleteDuplicates(a)
    while node:
        print node.val
        node = node.next
    print '----------------'
    a, b = ListNode(1), ListNode(2)
    c, d = ListNode(3), ListNode(3)
    e, f = ListNode(4), ListNode(4)
    g = ListNode(5)
    a.next, b.next = b, c
    c.next, d.next = d, e
    e.next, f.next = f, g
    node = solution.deleteDuplicates(a)
    while node:
        print node.val
        node = node.next

File: src/algorithms/cpp/Remove_Duplicates_from_Sorted_List_II.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        ListNode *pre_end = NULL, *pre = NULL, *beg = head, *node = head, *new_head = head;
        int cnt = 0;
        while (node) {
            if (node->val == beg->val) {
                ++ cnt;
            }
            else {
                if (cnt > 1) {
                    if (pre_end)    pre_end->next = node;
                    else            new_head = node;
                }
                else {
                    pre_end = pre;
                }
                beg = node;
                cnt = 1;
            }
            pre = node;
            node = node->next;
        }
        if (cnt > 1) {
            if (pre_end)    pre_end->next = node;
            else            new_head = node;
        }
        return new_head;
    }
};

int main() {
    ListNode *arr[7];
    arr[0] = new ListNode(1);
    arr[1] = new ListNode(1);
    arr[2] = new ListNode(3);
    arr[3] = new ListNode(3);
    arr[4] = new ListNode(4);
    arr[5] = new ListNode(4);
    arr[6] = new ListNode(5);
    for (int i = 0; i < 6; ++ i)
        arr[i]->next = arr[i + 1];
    Solution solution = Solution();
    ListNode *result = solution.deleteDuplicates(arr[0]);
    while (result) {
        cout << result->val << endl;
        result = result->next;
    }
    return 0;
}
